node/pkg/processor/broadcast.go

```diff
@@ -50,20 +49,20 @@ func (p *Processor) broadcastSignature(
 	// Store our VAA in case we're going to submit it to Solana
 	hash := hex.EncodeToString(digest.Bytes())
 
-	if p.state.signatures[hash] == nil {
-		p.state.signatures[hash] = &state{
-			firstObserved: time.Now(),
-			nextRetry:     time.Now().Add(nextRetryDuration(0)),
-			signatures:    map[ethcommon.Address][]byte{},
-			source:        "loopback",
-		}
+	state, created := p.state.getOrCreateState(hash)
+	state.lock.Lock()
+	defer state.lock.Unlock()
+
+	if created {
+		state.source = "loopback"
+		state.nextRetry = time.Now().Add(nextRetryDuration(0))
 	}
 
-	p.state.signatures[hash].ourObservation = o
-	p.state.signatures[hash].ourMsg = msg
-	p.state.signatures[hash].txHash = txhash
-	p.state.signatures[hash].source = o.GetEmitterChain().String()
-	p.state.signatures[hash].gs = p.gs // guaranteed to match ourObservation - there's no concurrent access to p.gs
+	state.ourObservation = o
+	state.ourMsg = msg
+	state.txHash = txhash
+	state.source = o.GetEmitterChain().String()
+	state.gs = p.gs.Load() // guaranteed to match ourObservation - there's no concurrent access to p.gs: TODO: Is this comment a problem??
```

Comment:
There will be concurrent access to p.gs now, so by the time we get here, p.gs may (in theory) be overwritten by a newer guardian set. So the assertion in the comment no longer holds. But the polling for guardian sets happens asynchronously from the message watching anyways, so I think this comment was never relevant. 
I wonder if `state.gs` is needed at all or if it would be sufficient to use `p.gs` everywhere.

Reply:
Had an offline discussion with @tbjump and @evan-gray and we agreed that it is not necessary to save the per-observation guardian set. I've upgraded the code to reflect this.